### DISCUSSION TITLE
Missing generic types for the `render` property

### DISCUSSION THREAD

**Discussion by bholmesdev** [markdoc/markdoc#400] [RFP]
The Astro team discovered a type safety issue once we allowed components to be passed to a tag's `render` property. We attempted a PR to markdoc core, but decided to step back and discuss as this proved complex.

## Background and motivation

Today, Astro lets users [apply components directly to the `render` property](https://docs.astro.build/en/guides/integrations-guide/markdoc/#use-astro-components-as-markdoc-tags) for Markdoc tags and nodes. This mirrors the config design of Markdoc's NextJS configuration, where you can import and apply JSX for rendering.

Using the default Markdoc config, you'll receive a type error. This is because `render` is of type `string` by default, and [the `Config` type does not expose a generic](https://github.com/markdoc/markdoc/blob/main/src/types.ts#L25) to change this.

We noticed the Schema type _does_ have such a generic:

```ts
export type Schema<C extends Config = Config, R = string> = {
  render?: R;
  ...
}
```
...Though this would require us to surgically update the Config like so:

```ts
export type AstroMarkdocConfi = Omit<
	MarkdocConfig,
	'tags' | 'nodes'
> &
	Partial<{
		tags: Record<string, Schema<Config, string | Component>>;
		nodes: Partial<Record<NodeType, Schema<Config, string | Component>>>;
	}>;
```

What's more, this generic does not pass `R` to the `C` generic for the `Config.` This can create problems if you need to check the config from a transform, and need to see if a `render` property is present. We do this today to conditionally check if [the user _has_ applied a `render` property](https://github.com/withastro/astro/blob/main/packages/integrations/markdoc/src/heading-ids.ts#L49) in our heading ID slugifier. `render` is forced to be a `string` here. 

## Attempted Solution

We started drafting an implementation PR that:
1. Introduces an `R` generic to the `Config` and `ConfigType`
2. (BREAKING) Reorders the `R` and `C` generics on the `Schema` type so render can be applied to the config correctly

https://github.com/markdoc/markdoc/compare/main...bholmesdev:markdoc:fix/type-defs

However, making `render` into a generic like this creates a cascade of type errors for all other types that rely on the `Config`. This includes the name attribute for `Tag` and the `transformAttributes()` and `transformChildren()` properties for `Node`. Refactoring these sites to use generics slowly caused generics to spread through the entire codebase 😬 

## Questions to address

Given the challenge of _truly_ keeping the `render` property generic, we want to answer:
- _Should_ `render` be a generic type accepting any value?
- Could Markdoc remove this generic and enforce `string | Function` as accepted properties? This would cover HTML elements, JSX functions, and Astro components.

**Comment by mfix-stripe** [comment]
I'm pretty peaceful with this direction, though I do have some hesitations around extending the generics types in the future if we need to.

Just to throw out an alternative, we could provide our own utility type that does the override you suggested above:
```ts
export type OverrideConfig<R, C extends Config = Config> = Omit<
  C,
  'tags' | 'nodes'
> &
  Partial<{
    tags: Record<string, Schema<C, R>>;
    nodes: Partial<Record<NodeType, Schema<C, R>>>;
  }>;
```

which would then be used like:
```ts
type OurConfig = OverrideConfig<string | Component>
```
which would avoid the breaking changes (though to be clear, I don't think this type of breaking change is too big of a deal before v1.0.0)

WDTY @rpaul-stripe
